server/gc/src/main/java/org/apache/accumulo/gc/GCRun.java

```diff
@@ -172,7 +177,19 @@ public Stream<Reference> getReferences() {
 
   @Override
   public Set<TableId> getTableIDs() {
-    return context.getTableIdToNameMap().keySet();
+    final String tablesPath = context.getZooKeeperRoot() + Constants.ZTABLES;
+    final ZooReader zr = context.getZooReader();
+    final Set<TableId> tids = new HashSet<>();
+    while (true) {
+      try {
+        zr.sync(tablesPath);
+        zr.getChildren(tablesPath).forEach(t -> tids.add(TableId.of(t)));
+        return tids;
+      } catch (KeeperException | InterruptedException e) {
+        log.error("Error getting tables from ZooKeeper, retrying", e);
+        UtilWaitThread.sleepUninterruptibly(1, TimeUnit.SECONDS);
+      }
+    }
```

Comment:
Maybe the new `getCandidateTableIDs()` method is not needed and this method could be modified.  Looking around at the code there is only one place this is used currently and I think this change would be good for that code.   The existing code that uses it may not be correct for metadata and root  levels (but I think luckily it may not matter).

```suggestion
    switch(level){
    case ROOT:
        return Set.of(RootTable.ID);
    case METADATA:
        return Set.of(MetadataTable.ID);
    case USER:
      final String tablesPath = context.getZooKeeperRoot() + Constants.ZTABLES;
      final ZooReader zr = context.getZooReader();
      final Set<TableId> tids = new HashSet<>();
      while (true) {
        try {
          zr.sync(tablesPath);
          zr.getChildren(tablesPath).forEach(t -> tids.add(TableId.of(t)));
          return tids;
        } catch (KeeperException | InterruptedException e) {
          log.error("Error getting tables from ZooKeeper, retrying", e);
          UtilWaitThread.sleepUninterruptibly(1, TimeUnit.SECONDS);
        }
      }
    default:
       throw new IllegalArgumentException("unknown level "+level);
    }
```

Reply:
> The existing code that uses it may not be correct for metadata and root levels 

Actually the existing code that uses it is probably correct if the current impl returns all table ids including root and metadata tables which I think it would as long as those are in ZK.  Limiting it to returning just the table ids for the current level should not harm the current usage.

Reply:
Maybe it makes sense to have two methods for clarity though.